src/storage/v3/result.hpp

```diff
@@ -11,24 +11,36 @@
 
 #pragma once
 
+#include <cstdint>
+#include <string>
+#include <string_view>
 #include <type_traits>
 
+#include "common/errors.hpp"
 #include "utils/result.hpp"
 
 namespace memgraph::storage::v3 {
 
 static_assert(std::is_same_v<uint8_t, unsigned char>);
 
-enum class Error : uint8_t {
-  SERIALIZATION_ERROR,
-  NONEXISTENT_OBJECT,
-  DELETED_OBJECT,
-  VERTEX_HAS_EDGES,
-  PROPERTIES_DISABLED,
-  VERTEX_ALREADY_INSERTED
+struct ShardError {
+  ShardError(common::ErrorCode code, std::string message, std::string source)
+      : code{code}, message{std::move(message)}, source{std::move(source)} {}
+
+  ShardError(common::ErrorCode code, std::string source) : code{code}, source{std::move(source)} {}
+
+  common::ErrorCode code;
+  // TODO Maybe add category
+  std::string message;
+  std::string source;
+
+  inline friend bool operator==(const ShardError &lhs, const ShardError &rhs) { return lhs.code == rhs.code; }
 };
 
+// NOLINTNEXTLINE(cppcoreguidelines-macro-usage)
+#define SHARD_ERROR(...) memgraph::storage::v3::ShardError(__VA_ARGS__, fmt::format("{}:{}", __FILE__, __LINE__))
```

Comment:
Maybe use (`std::source_location`)[https://en.cppreference.com/w/cpp/experimental/source_location]? We can store that also in the error. I know I mentioned the `std::string` source, but I just realized that now `source_location` is a superior option. Sorry for that.

Reply:
Try `<source_location>`, maybe it works because we have stdlibc++ 11 which officially supports it.

Reply:
`https://en.cppreference.com/w/cpp/compiler_support/20` we have a lib support but not a compiler, it is supported by clang 15 and greater